File: .windsurf/scripts/backup_database.dart
#!/usr/bin/env dart
// ignore_for_file: avoid_print

import 'dart:convert';
import 'dart:io';
import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:firebase_core/firebase_core.dart';

/// 🔐 FitX Database Backup Script
/// 
/// هذا السكريبت بيعمل أرشفة كاملة لكل بيانات التطبيق من Firebase Firestore
/// 
/// الاستخدام:
/// ```bash
/// dart backup_database.dart
/// ```
///
/// المخرجات:
/// - backup_YYYY-MM-DD/          (مجلد بتاريخ النسخة)
///   ├── users/                  (كل بيانات المستخدمين)
///   ├── linkRequests/          (طلبات الربط)
///   ├── tasks/                 (المهام)
///   ├── workouts/              (التمرينات)
///   ├── exercises/             (التمارين)
///   └── metadata.json          (معلومات النسخة)

void main() async {
  print('🚀 FitX Database Backup Script');
  print('═══════════════════════════════════════════════════════════════');
  
  final startTime = DateTime.now();
  final timestamp = startTime.toIso8601String().split('T')[0];
  final backupDir = Directory('backup_$timestamp');
  
  // Initialize Firebase
  print('🔌 Connecting to Firebase...');
  await Firebase.initializeApp(
    options: const FirebaseOptions(
      apiKey: 'YOUR_API_KEY',
      appId: 'YOUR_APP_ID',
      messagingSenderId: 'YOUR_SENDER_ID',
      projectId: 'fitx-app',
    ),
  );
  
  final firestore = FirebaseFirestore.instance;
  
  // Create backup directory
  if (!backupDir.existsSync()) {
    backupDir.createSync(recursive: true);
  }
  
  print('📁 Backup directory: ${backupDir.path}');
  print('');
  
  // Backup statistics
  final stats = <String, int>{};
  
  try {
    // ═══════════════════════════════════════════════════════════════
    // 1. BACKUP USERS (Most Critical)
    // ═══════════════════════════════════════════════════════════════
    print('👥 Backing up USERS collection...');
    final usersDir = Directory('${backupDir.path}/users');
    usersDir.createSync();
    
    final usersSnapshot = await firestore.collection('users').get();
    final userCount = await _backupUsers(firestore, usersDir, usersSnapshot);
    stats['users'] = userCount;
    print('   ✅ $userCount users backed up');
    
    // ═══════════════════════════════════════════════════════════════
    // 2. BACKUP LINK REQUESTS
    // ═══════════════════════════════════════════════════════════════
    print('🔗 Backing up LINK REQUESTS...');
    final linkRequestsDir = Directory('${backupDir.path}/linkRequests');
    linkRequestsDir.createSync();
    
    final linkRequestsSnapshot = await firestore.collection('linkRequests').get();
    final linkCount = await _backupCollection(
      linkRequestsSnapshot, 
      linkRequestsDir, 
      'linkRequests'
    );
    stats['linkRequests'] = linkCount;
    print('   ✅ $linkCount link requests backed up');
    
    // ═══════════════════════════════════════════════════════════════
    // 3. BACKUP TASKS
    // ═══════════════════════════════════════════════════════════════
    print('📋 Backing up TASKS...');
    final tasksDir = Directory('${backupDir.path}/tasks');
    tasksDir.createSync();
    
    final tasksSnapshot = await firestore.collection('tasks').get();
    final taskCount = await _backupCollection(tasksSnapshot, tasksDir, 'tasks');
    stats['tasks'] = taskCount;
    print('   ✅ $taskCount tasks backed up');
    
    // ═══════════════════════════════════════════════════════════════
    // 4. BACKUP WORKOUTS
    // ═══════════════════════════════════════════════════════════════
    print('💪 Backing up WORKOUTS...');
    final workoutsDir = Directory('${backupDir.path}/workouts');
    workoutsDir.createSync();
    
    final workoutsSnapshot = await firestore.collection('workouts').get();
    final workoutCount = await _backupCollection(workoutsSnapshot, workoutsDir, 'workouts');
    stats['workouts'] = workoutCount;
    print('   ✅ $workoutCount workouts backed up');
    
    // ═══════════════════════════════════════════════════════════════
    // 5. BACKUP EXERCISES
    // ═══════════════════════════════════════════════════════════════
    print('🏋️ Backing up EXERCISES...');
    final exercisesDir = Directory('${backupDir.path}/exercises');
    exercisesDir.createSync();
    
    final exercisesSnapshot = await firestore.collection('exercises').get();
    final exerciseCount = await _backupCollection(exercisesSnapshot, exercisesDir, 'exercises');
    stats['exercises'] = exerciseCount;
    print('   ✅ $exerciseCount exercises backed up');
    
    // ═══════════════════════════════════════════════════════════════
    // 6. BACKUP MUSCLE GROUPS & ANGLES
    // ═══════════════════════════════════════════════════════════════
    print('🦾 Backing up MUSCLE GROUPS...');
    final muscleGroupsDir = Directory('${backupDir.path}/muscleGroups');
    muscleGroupsDir.createSync();
    
    final muscleGroupsSnapshot = await firestore.collection('muscleGroups').get();
    final muscleCount = await _backupCollection(
      muscleGroupsSnapshot, 
      muscleGroupsDir, 
      'muscleGroups'
    );
    stats['muscleGroups'] = muscleCount;
    print('   ✅ $muscleCount muscle groups backed up');
    
    // ═══════════════════════════════════════════════════════════════
    // 7. BACKUP NUTRITION DATA
    // ═══════════════════════════════════════════════════════════════
    print('🥗 Backing up NUTRITION...');
    final nutritionDir = Directory('${backupDir.path}/nutrition');
    nutritionDir.createSync();
    
    final foodItemsSnapshot = await firestore.collection('foodItems').get();
    final mealLogsSnapshot = await firestore.collection('mealLogs').get();
    
    final foodCount = await _backupCollection(foodItemsSnapshot, nutritionDir, 'foodItems');
    final mealCount = await _backupCollection(mealLogsSnapshot, nutritionDir, 'mealLogs');
    
    stats['foodItems'] = foodCount;
    stats['mealLogs'] = mealCount;
    print('   ✅ $foodCount food items backed up');
    print('   ✅ $mealCount meal logs backed up');
    
    // ═══════════════════════════════════════════════════════════════
    // 8. CREATE METADATA FILE
    // ═══════════════════════════════════════════════════════════════
    print('📝 Creating metadata...');
    final endTime = DateTime.now();
    final duration = endTime.difference(startTime);
    
    final metadata = {
      'backupDate': startTime.toIso8601String(),
      'completedAt': endTime.toIso8601String(),
      'durationSeconds': duration.inSeconds,
      'statistics': stats,
      'totalDocuments': stats.values.reduce((a, b) => a + b),
      'firebaseProject': 'fitx-app',
      'backupVersion': '1.0',
    };
    
    final metadataFile = File('${backupDir.path}/metadata.json');
    await metadataFile.writeAsString(
      const JsonEncoder.withIndent('  ').convert(metadata)
    );
    
    // ═══════════════════════════════════════════════════════════════
    // 9. CREATE SUMMARY REPORT
    // ═══════════════════════════════════════════════════════════════
    print('');
    print('═══════════════════════════════════════════════════════════════');
    print('✅ BACKUP COMPLETED SUCCESSFULLY!');
    print('═══════════════════════════════════════════════════════════════');
    print('');
    print('📊 Summary:');
    print('   Duration: ${duration.inSeconds} seconds');
    print('   Total documents: ${stats.values.reduce((a, b) => a + b)}');
    print('');
    print('📁 Collections backed up:');
    stats.forEach((collection, count) {
      print('   • $collection: $count documents');
    });
    print('');
    print('💾 Backup location: ${backupDir.absolute.path}');
    print('');
    print('🔐 SECURITY NOTE:');
    print('   • Keep this backup secure and encrypted');
    print('   • Do not commit to Git');
    print('   • Store in multiple locations');
    print('');
    
  } catch (e, stackTrace) {
    print('');
    print('❌ BACKUP FAILED!');
    print('Error: $e');
    print('Stack trace: $stackTrace');
    exit(1);
  }
}

/// Backup users with their subcollections
Future<int> _backupUsers(
  FirebaseFirestore firestore, 
  Directory usersDir, 
  QuerySnapshot usersSnapshot
) async {
  int count = 0;
  
  for (final userDoc in usersSnapshot.docs) {
    final userId = userDoc.id;
    final userDir = Directory('${usersDir.path}/$userId');
    userDir.createSync();
    
    // Save user main data
    final userData = _sanitizeDocumentData(userDoc.data() as Map<String, dynamic>);
    userData['_docId'] = userId;
    userData['_backupTime'] = DateTime.now().toIso8601String();
    
    await File('${userDir.path}/profile.json').writeAsString(
      const JsonEncoder.withIndent('  ').convert(userData)
    );
    
    // Backup subcollections
    await _backupSubcollection(
      firestore, userId, 'daily_stats', '${userDir.path}/daily_stats'
    );
    await _backupSubcollection(
      firestore, userId, 'exerciseHistory', '${userDir.path}/exerciseHistory'
    );
    await _backupSubcollection(
      firestore, userId, 'exerciseStats', '${userDir.path}/exerciseStats'
    );
    await _backupSubcollection(
      firestore, userId, 'progressPhotos', '${userDir.path}/progressPhotos'
    );
    await _backupSubcollection(
      firestore, userId, 'nutritionLogs', '${userDir.path}/nutritionLogs'
    );
    
    count++;
    
    // Progress indicator every 10 users
    if (count % 10 == 0) {
      print('   ... backed up $count users');
    }
  }
  
  return count;
}

/// Backup a single collection
Future<int> _backupCollection(
  QuerySnapshot snapshot, 
  Directory dir, 
  String collectionName
) async {
  int count = 0;
  final batchData = <Map<String, dynamic>>[];
  
  for (final doc in snapshot.docs) {
    final data = _sanitizeDocumentData(doc.data() as Map<String, dynamic>);
    data['_docId'] = doc.id;
    data['_backupTime'] = DateTime.now().toIso8601String();
    batchData.add(data);
    count++;
  }
  
  // Save as single JSON file
  final file = File('${dir.path}/$collectionName.json');
  await file.writeAsString(
    const JsonEncoder.withIndent('  ').convert(batchData)
  );
  
  return count;
}

/// Backup a subcollection for a specific user
Future<int> _backupSubcollection(
  FirebaseFirestore firestore,
  String userId,
  String subcollectionName,
  String outputPath,
) async {
  try {
    final snapshot = await firestore
        .collection('users')
        .doc(userId)
        .collection(subcollectionName)
        .get();
    
    if (snapshot.docs.isEmpty) return 0;
    
    final dir = Directory(outputPath);
    dir.createSync(recursive: true);
    
    final batchData = <Map<String, dynamic>>[];
    
    for (final doc in snapshot.docs) {
      final data = _sanitizeDocumentData(doc.data());
      data['_docId'] = doc.id;
      data['_backupTime'] = DateTime.now().toIso8601String();
      batchData.add(data);
    }
    
    final file = File('$outputPath.json');
    await file.writeAsString(
      const JsonEncoder.withIndent('  ').convert(batchData)
    );
    
    return snapshot.docs.length;
  } catch (e) {
    print('   ⚠️ Warning: Could not backup $subcollectionName for user $userId: $e');
    return 0;
  }
}

/// Sanitize document data for JSON serialization
/// Converts Timestamps to ISO strings
Map<String, dynamic> _sanitizeDocumentData(Map<String, dynamic> data) {
  final sanitized = <String, dynamic>{};
  
  data.forEach((key, value) {
    if (value is DateTime) {
      sanitized[key] = value.toIso8601String();
    } else if (value is Map<String, dynamic>) {
      sanitized[key] = _sanitizeDocumentData(value);
    } else if (value is List) {
      sanitized[key] = value.map((item) {
        if (item is Map<String, dynamic>) {
          return _sanitizeDocumentData(item);
        } else if (item is DateTime) {
          return item.toIso8601String();
        }
        return item;
      }).toList();
    } else {
      sanitized[key] = value;
    }
  });
  
  return sanitized;
}

File: .windsurf/scripts/backup_database.js
#!/usr/bin/env node

/**
 * 🔐 FitX Database Backup Script (Node.js Version)
 * 
 * هذا السكريبت بيعمل أرشفة كاملة لكل بيانات التطبيق من Firebase Firestore
 * 
 * المتطلبات:
 * npm install firebase-admin
 * 
 * الاستخدام:
 * node backup_database.js
 * 
 * أو مع Service Account:
 * node backup_database.js --service-account=path/to/serviceAccount.json
 */

const admin = require('firebase-admin');
const fs = require('fs');
const path = require('path');

// Initialize Firebase Admin
// You need to download serviceAccountKey.json from Firebase Console
// Project Settings > Service Accounts > Generate New Private Key
const serviceAccount = require('./serviceAccountKey.json');

admin.initializeApp({
  credential: admin.credential.cert(serviceAccount),
  databaseURL: `https://${serviceAccount.project_id}.firebaseio.com`
});

const db = admin.firestore();

// Collections to backup
const COLLECTIONS = [
  'users',
  'linkRequests',
  'tasks',
  'workouts',
  'exercises',
  'muscleGroups',
  'foodItems',
  'mealLogs',
  'notifications',
];

// User subcollections to backup
const USER_SUBCOLLECTIONS = [
  'daily_stats',
  'exerciseHistory',
  'exerciseStats',
  'progressPhotos',
  'nutritionLogs',
  'tasks',
];

async function backupDatabase() {
  const startTime = new Date();
  const timestamp = startTime.toISOString().split('T')[0];
  const backupDir = path.join(__dirname, `backup_${timestamp}`);
  
  console.log('🚀 FitX Database Backup Script (Node.js)');
  console.log('═══════════════════════════════════════════════════════════════');
  console.log(`📁 Backup directory: ${backupDir}`);
  console.log('');
  
  // Create backup directory
  if (!fs.existsSync(backupDir)) {
    fs.mkdirSync(backupDir, { recursive: true });
  }
  
  const stats = {};
  
  try {
    // ═══════════════════════════════════════════════════════════════
    // BACKUP MAIN COLLECTIONS
    // ═══════════════════════════════════════════════════════════════
    for (const collectionName of COLLECTIONS) {
      console.log(`📦 Backing up ${collectionName}...`);
      
      const collectionDir = path.join(backupDir, collectionName);
      if (!fs.existsSync(collectionDir)) {
        fs.mkdirSync(collectionDir);
      }
      
      const snapshot = await db.collection(collectionName).get();
      const documents = [];
      
      snapshot.forEach(doc => {
        documents.push({
          _docId: doc.id,
          ...sanitizeData(doc.data()),
          _backupTime: new Date().toISOString()
        });
      });
      
      // Save as JSON
      const filePath = path.join(collectionDir, `${collectionName}.json`);
      fs.writeFileSync(filePath, JSON.stringify(documents, null, 2));
      
      stats[collectionName] = documents.length;
      console.log(`   ✅ ${documents.length} documents backed up`);
      
      // For users collection, also backup subcollections
      if (collectionName === 'users') {
        console.log('   👥 Backing up user subcollections...');
        let subcollectionCount = 0;
        
        for (const doc of snapshot.docs) {
          const userId = doc.id;
          const userDir = path.join(collectionDir, userId);
          fs.mkdirSync(userDir, { recursive: true });
          
          for (const subcollection of USER_SUBCOLLECTIONS) {
            const subSnapshot = await db
              .collection('users')
              .doc(userId)
              .collection(subcollection)
              .get();
            
            if (subSnapshot.empty) continue;
            
            const subDocs = [];
            subSnapshot.forEach(subDoc => {
              subDocs.push({
                _docId: subDoc.id,
                ...sanitizeData(subDoc.data()),
                _backupTime: new Date().toISOString()
              });
            });
            
            const subFilePath = path.join(userDir, `${subcollection}.json`);
            fs.writeFileSync(subFilePath, JSON.stringify(subDocs, null, 2));
            subcollectionCount += subDocs.length;
          }
        }
        
        console.log(`   ✅ ${subcollectionCount} subcollection documents backed up`);
      }
    }
    
    // ═══════════════════════════════════════════════════════════════
    // CREATE METADATA
    // ═══════════════════════════════════════════════════════════════
    const endTime = new Date();
    const duration = (endTime - startTime) / 1000;
    
    const totalDocuments = Object.values(stats).reduce((a, b) => a + b, 0);
    
    const metadata = {
      backupDate: startTime.toISOString(),
      completedAt: endTime.toISOString(),
      durationSeconds: duration,
      statistics: stats,
      totalDocuments: totalDocuments,
      firebaseProject: serviceAccount.project_id,
      backupVersion: '1.0',
      backupTool: 'FitX Node.js Backup Script'
    };
    
    const metadataPath = path.join(backupDir, 'metadata.json');
    fs.writeFileSync(metadataPath, JSON.stringify(metadata, null, 2));
    
    // ═══════════════════════════════════════════════════════════════
    // CREATE HUMAN-READABLE REPORT
    // ═══════════════════════════════════════════════════════════════
    const reportPath = path.join(backupDir, 'REPORT.txt');
    const report = `
╔════════════════════════════════════════════════════════════════╗
║           FitX Database Backup Report                          ║
╚════════════════════════════════════════════════════════════════╝

Backup Date: ${startTime.toLocaleString('ar-EG')}
Project: ${serviceAccount.project_id}
Duration: ${duration.toFixed(2)} seconds

SUMMARY
───────
Total Documents: ${totalDocuments}
Collections: ${Object.keys(stats).length}

DETAILS
───────
${Object.entries(stats).map(([name, count]) => `  • ${name}: ${count} documents`).join('\n')}

SECURITY NOTES
──────────────
• Keep this backup secure and encrypted
• Do not commit to version control
• Store in multiple secure locations
• Verify backup integrity periodically

RESTORE INSTRUCTIONS
────────────────────
To restore from this backup, use the restore script:
  node restore_database.js backup_${timestamp}

Or manually import JSON files to Firebase using:
  firebase firestore:delete --all-collections
  firebase firestore:import backup_${timestamp}

Generated by: FitX Backup Script v1.0
`;
    
    fs.writeFileSync(reportPath, report);
    
    // ═══════════════════════════════════════════════════════════════
    // FINAL OUTPUT
    // ═══════════════════════════════════════════════════════════════
    console.log('');
    console.log('═══════════════════════════════════════════════════════════════');
    console.log('✅ BACKUP COMPLETED SUCCESSFULLY!');
    console.log('═══════════════════════════════════════════════════════════════');
    console.log('');
    console.log(`📊 Summary:`);
    console.log(`   Duration: ${duration.toFixed(2)} seconds`);
    console.log(`   Total documents: ${totalDocuments}`);
    console.log('');
    console.log('📁 Collections backed up:');
    Object.entries(stats).forEach(([name, count]) => {
      console.log(`   • ${name}: ${count} documents`);
    });
    console.log('');
    console.log(`💾 Backup location: ${path.resolve(backupDir)}`);
    console.log(`📝 Report file: ${path.join(backupDir, 'REPORT.txt')}`);
    console.log('');
    console.log('🔐 SECURITY NOTE:');
    console.log('   • Keep this backup secure and encrypted');
    console.log('   • Do not commit to Git');
    console.log('   • Store in multiple locations');
    console.log('');
    
  } catch (error) {
    console.error('');
    console.error('❌ BACKUP FAILED!');
    console.error('Error:', error.message);
    console.error('Stack:', error.stack);
    process.exit(1);
  }
}

/**
 * Sanitize data for JSON serialization
 * Converts Firestore timestamps to ISO strings
 */
function sanitizeData(data) {
  if (data === null || data === undefined) {
    return null;
  }
  
  if (data instanceof Date) {
    return data.toISOString();
  }
  
  if (data._seconds !== undefined && data._nanoseconds !== undefined) {
    // Firestore Timestamp
    return new Date(data._seconds * 1000).toISOString();
  }
  
  if (Array.isArray(data)) {
    return data.map(item => sanitizeData(item));
  }
  
  if (typeof data === 'object') {
    const sanitized = {};
    for (const [key, value] of Object.entries(data)) {
      sanitized[key] = sanitizeData(value);
    }
    return sanitized;
  }
  
  return data;
}

// Run backup
backupDatabase();
